Framework/Algorithms/src/CopySample.cpp

```diff
@@ -150,6 +158,52 @@ void CopySample::copyParameters(Sample &from, Sample &to, bool nameFlag, bool ma
       rhsMaterial = to.getMaterial();
     }
     auto rhsObject = std::shared_ptr<IObject>(from.getShape().cloneWithMaterial(rhsMaterial));
+
+    if (auto csgObj = std::dynamic_pointer_cast<Geometry::CSGObject>(rhsObject)) {
+      // Rotate CSGObject by goniometer by editing XML, if possible for that shape
+
+      std::string xml = rhsObject->getShapeXML();
+
+      // Delete previous goniometer from xml
+      std::size_t foundGonioTag = xml.find("<goniometer");
+      if (foundGonioTag != std::string::npos) {
+        std::size_t gonioTagLength = xml.find(">", foundGonioTag + 1) - foundGonioTag;
+        xml.erase(foundGonioTag, gonioTagLength);
+      }
+
+      // Put goniometer tag in correct place in xml
+      std::size_t gonioPlace;
+      std::size_t foundAlgebra = xml.find("<algebra");
+      std::size_t foundAlgebraEnd = xml.find("/>", foundAlgebra) + 2;
+      std::size_t foundType = xml.find("</type>");
+
+      if (foundAlgebra != std::string::npos) {
+        // If Algebra tag exists, add goniometer AFTER algebra end tag
+        gonioPlace = foundAlgebraEnd;
+      } else if (foundType != std::string::npos) {
+        // If no algebra tag, add goniometer BEFORE Type end tag
+        gonioPlace = foundType;
+      } else {
+        // If no Algebra or Type tag, add goniometer to the end
+        gonioPlace = xml.size();
+      }
+
+      const std::vector<std::string> matrixElementNames = {"a11", "a12", "a13", "a21", "a22",
+                                                           "a23", "a31", "a32", "a33"};
+      std::string goniometerRotation = " <goniometer ";
+      for (size_t index = 0; index < rotationMatrix.size(); ++index) {
+        goniometerRotation += matrixElementNames[index] + " = '" + std::to_string(rotationMatrix[index]) + "' ";
+      }
+      goniometerRotation += "/>";
+      xml.insert(gonioPlace, goniometerRotation);
```

Comment:
The 10-20 lines here is also in the SetSample.cpp unit. Perhaps this could be moved into the ShapeFactor as a new method? The method could take the rotation matrix and the shape xml as parameters perhaps and insert the goniometer tag